File: frontend/src/popup/composables/cardConfigSync.ts
import type { AppConfig } from '@/types/config'
import type { CardState } from './useTranslationEvents'

export type EnabledServicePayload = {
  serviceInstanceId: string
  serviceType: string
  serviceName: string
  protocol: string
  modelName: string
}

/** 按 config.services 数组顺序取已启用服务，供结果卡排序与元数据同步。 */
export function enabledPayloads(config: AppConfig): EnabledServicePayload[] {
  return (config.services || [])
    .filter((s) => s.enabled)
    .map((s) => ({
      serviceInstanceId: s.id,
      serviceType: s.serviceType,
      serviceName: s.name,
      protocol: s.protocol || '',
      modelName: s.protocol === 'microsoft_edge' ? '' : s.model || '',
    }))
}

function createPendingCard(p: EnabledServicePayload): CardState {
  return {
    serviceInstanceId: p.serviceInstanceId,
    serviceName: p.serviceName,
    serviceType: p.serviceType,
    protocol: p.protocol,
    modelName: p.modelName,
    text: '',
    status: 'pending',
    collapsed: true,
    collapseUserOverride: false,
    expanded: false,
    hasOverflow: false,
    showActions: false,
    usage: null,
    detectedSourceLang: null,
    errorTitleKey: null,
    errorMessage: '',
  }
}

function applyMeta(card: CardState, p: EnabledServicePayload): void {
  card.serviceName = p.serviceName
  card.serviceType = p.serviceType
  card.protocol = p.protocol
  card.modelName = p.modelName
}

/**
 * 按启用服务列表顺序重建 cards Map。
 * - 空闲：增删改卡片，Map 迭代序 = 配置启用序
 * - 翻译中：不新增卡片；可删非 translating 的已禁用卡；元数据可更新；
 *   仍保留的卡按配置序重排（翻译中新启用不参与当前批次）
 */
export function syncCardsFromEnabledServices(
  cards: Map<string, CardState>,
  payloads: EnabledServicePayload[],
  options: { isTranslating: boolean },
): void {
  const next = new Map<string, CardState>()

  for (const p of payloads) {
    const existing = cards.get(p.serviceInstanceId)
    if (existing) {
      applyMeta(existing, p)
      next.set(p.serviceInstanceId, existing)
    } else if (!options.isTranslating) {
      next.set(p.serviceInstanceId, createPendingCard(p))
    }
  }

  // 翻译中：保留仍在输出、但已从启用列表移除的卡（挂在末尾）
  if (options.isTranslating) {
    for (const [id, card] of cards) {
      if (!next.has(id) && card.status === 'translating') {
        next.set(id, card)
      }
    }
  }

  cards.clear()
  for (const [id, card] of next) {
    cards.set(id, card)
  }
}

File: native/windows/popup/State/CardConfigSync.cs
namespace Shizi.Popup.State;

/// <summary>
/// 按启用服务列表同步卡片 Map，对齐 Vue <c>syncCardsFromEnabledServices</c>。
/// </summary>
public static class CardConfigSync
{
    /// <summary>
    /// - 空闲：按启用序增删改卡片
    /// - 翻译中：不新增未参与服务卡；可删非 translating 的已禁用卡；元数据可更新；
    ///   仍在 translating 但已禁用的卡挂在末尾
    /// </summary>
    public static void SyncCards(
        IDictionary<string, CardState> cards,
        IReadOnlyList<EnabledService> enabled,
        bool isTranslating)
    {
        var next = new Dictionary<string, CardState>();

        foreach (var p in enabled)
        {
            if (cards.TryGetValue(p.ServiceInstanceId, out var existing))
            {
                ApplyMeta(existing, p);
                next[p.ServiceInstanceId] = existing;
            }
            else if (!isTranslating)
            {
                next[p.ServiceInstanceId] = CardState.CreatePending(
                    p.ServiceInstanceId,
                    p.ServiceName,
                    p.ServiceType,
                    p.Protocol,
                    p.ModelName);
            }
        }

        // 翻译中：保留仍在输出、但已从启用列表移除的卡（挂在末尾）
        if (isTranslating)
        {
            foreach (var (id, card) in cards)
            {
                if (!next.ContainsKey(id) && card.Status == CardStatus.Translating)
                    next[id] = card;
            }
        }

        cards.Clear();
        foreach (var (id, card) in next)
            cards[id] = card;
    }

    /// <summary>
    /// 从完整服务配置中提取启用项（保序）。
    /// microsoft_edge 清空 modelName，对齐 Vue <c>enabledPayloads</c>。
    /// </summary>
    public static IReadOnlyList<EnabledService> EnabledPayloads(
        IEnumerable<(string Id, string Name, string ServiceType, string Protocol, string Model, bool Enabled)> services)
    {
        var list = new List<EnabledService>();
        foreach (var s in services)
        {
            if (!s.Enabled)
                continue;

            var modelName = s.Protocol == "microsoft_edge" ? "" : s.Model;
            list.Add(new EnabledService(
                s.Id,
                s.Name,
                s.ServiceType,
                s.Protocol,
                modelName));
        }

        return list;
    }

    private static void ApplyMeta(CardState card, EnabledService p)
    {
        card.ServiceName = p.ServiceName;
        card.ServiceType = p.ServiceType;
        card.Protocol = p.Protocol;
        card.ModelName = p.ModelName;
    }
}
